packages/shared/src/hooks/bookmark/useBookmarkFolderList.ts

```diff
@@ -1,23 +1,55 @@
-import { useQuery } from '@tanstack/react-query';
-import { BookmarkFolder, getBookmarkFolders } from '../../graphql/bookmarks';
+import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
+import { useCallback } from 'react';
+import {
+  BookmarkFolder,
+  getBookmarkFolders,
+  type MoveBookmarkProps,
+  moveBookmark as moveBookmarkMutation,
+} from '../../graphql/bookmarks';
 import { generateQueryKey, RequestKey, StaleTime } from '../../lib/query';
 import { useAuthContext } from '../../contexts/AuthContext';
+import { getPostByIdKey } from '../usePostById';
+import { useToastNotification } from '../useToastNotification';
 
 interface UseBookmarkFolderList {
   isPending: boolean;
   folders: Array<BookmarkFolder>;
+  isMoving: boolean;
+  onMoveBookmark: (props: MoveBookmarkProps) => void;
 }
 
 export const useBookmarkFolderList = (): UseBookmarkFolderList => {
+  const client = useQueryClient();
+  const { displayToast } = useToastNotification();
   const { isAuthReady, isLoggedIn } = useAuthContext();
   const { data, isPending } = useQuery({
     queryKey: generateQueryKey(RequestKey.BookmarkFolders),
     queryFn: getBookmarkFolders,
     staleTime: StaleTime.Default,
     enabled: isAuthReady && isLoggedIn,
   });
+  const { mutate: moveBookmark, isPending: isMoving } = useMutation({
+    mutationFn: moveBookmarkMutation,
+    onError: () => {
+      displayToast('❌ Failed to move bookmark');
+    },
+    onSuccess: (_, vars: MoveBookmarkProps) => {
+      client.invalidateQueries({
+        queryKey: getPostByIdKey(vars.postId),
+      });
```

Comment:
But you have all the data here for a optimistic update, we should actually have some handler already to optimistic update posts.
It's just preferred vs simply invalidating.

Reply:
I agree, normally we try to manually update the UI optimistically and only do invalidation when it is not possible.